### ISSUE TITLE
Declare all dependencies prior to making them available.

### ISSUE THREAD

**Issue by vkalintiris** [netdata/netdata#19076]
#### Summary

When we are using CMake's `FetchContent`, we are making available the dependency at the same time. This will lead to dependencies not being aware of other projects that rely on the same dependency. IOW, we need to make available the populated projects at once.

From [CMake's](https://cmake.org/cmake/help/v3.31/module/FetchContent.html#command:fetchcontent_makeavailable:~:text=Projects%20should%20aim,a%20chance%20to\).) documentation:

> Projects should aim to declare the details of all dependencies they might use before they call FetchContent_MakeAvailable() for any of them. This ensures that if any of the dependencies are also sub-dependencies of one or more of the others, the main project still controls the details that will be used (because it will declare them first before the dependencies get a chance to).

**Comment by yuvashrikarunakaran** [comment]
From CMake's Documentation:
CMake’s FetchContent module allows for the downloading and building of external dependencies within the same build process. Here’s an outline of how it typically works:

Basic FetchContent Usage: FetchContent allows you to specify dependencies that should be fetched from remote repositories and used in your build. Once the content is fetched, it is made available in the current CMake context.

cmake
Copy code
include(FetchContent)
FetchContent_Declare(
    SomeDependency
    GIT_REPOSITORY https://github.com/some/repo.git
    GIT_TAG v1.0.0
)
FetchContent_MakeAvailable(SomeDependency)
Issue of Multiple Dependencies: When multiple projects or modules within the same CMake build rely on the same dependency (but perhaps different versions or configurations), FetchContent does not natively handle the notion of "shared" dependencies well. This leads to each project fetching and building its own version of the dependency without awareness of other projects.

The Solution:
To address the issue where dependencies are not aware of each other, CMake provides a few strategies to ensure that dependencies are handled properly and shared across multiple projects.

1. Use FetchContent with EXCLUDE_FROM_ALL:
You can make the dependency available but prevent it from automatically being added to the build. This can be useful to make sure that it doesn't interfere with other targets that rely on the same dependency.

cmake
Copy code
FetchContent_Declare(
    SomeDependency
    GIT_REPOSITORY https://github.com/some/repo.git
    GIT_TAG v1.0.0
)
FetchContent_MakeAvailable(SomeDependency)

# Now, you can add your own dependencies that rely on SomeDependency, but it's not
# automatically added to the target.
2. Force a Shared Dependency across Multiple Projects:
You can ensure that a dependency is only fetched once by using FetchContent but managing it manually. This involves fetching the content once and passing the already populated content to different modules or projects.

cmake
Copy code
# Declare the dependency once
FetchContent_Declare(
    SomeDependency
    GIT_REPOSITORY https://github.com/some/repo.git
    GIT_TAG v1.0.0
)

# Populate it once
FetchContent_MakeAvailable(SomeDependency)

# Then use it across different projects or targets
add_subdirectory(OtherProject)
This ensures that the dependency is populated only once, and each project can use it as needed.

3. Use ExternalProject:
If you need more control over the fetching process, you could switch to using ExternalProject to manage the fetching and building of dependencies. This gives you finer control over how and when dependencies are built and avoids potential issues with FetchContent.

cmake
Copy code
include(ExternalProject)
ExternalProject_Add(
    SomeDependency
    GIT_REPOSITORY https://github.com/some/repo.git
    GIT_TAG v1.0.0
    PREFIX ${CMAKE_BINARY_DIR}/_deps
)
With this, you can control the build process more tightly and ensure that dependencies aren't fetched multiple times.

**Comment by Ferroin** [comment]
Did a bit of digging WRT this, and I’ve run into an issue.

TLDR: We need CMake 3.18 to do this cleanly, which means we either have to wait for Ubuntu 20.04 to go EOL in April, or we need to add some extra complexity to the CI to handle it cleanly.

Full explanation:

For various reasons, we need to run some extra code (mostly setting variables, but sometimes defining targets and/or custom commands) _after_ fetching certain dependencies that are handled through FetchContent, but this code needs to run _only_ if the dependency was actually handled through FetchContent (as a concrete example, for JSON-C we need to create a symlink at build time to make C/C++ includes work correctly for it, but only if it’s being handled by FetchContent).

In our current setup, this is a non-issue, because we fetch the individual dependencies as part of the functions that check for them and can thus just run the required extra code in the code path in those functions that fetches them.

If we centralize everything though so that we fetch everything only after all dependencies have been declared, we then need to also track what needs to be run as post-fetch code. In theory, we could do this by writing out a module at configure time and then including it later in the configuration process just after all the dependencies have been fetched, but that has a number of issues (not least of which being that generating CMake code from CMake code is tricky and error-prone due to CMake’s inability to escape `$` in a string and have variables evaluated in that same string). Alternatively, CMake 3.18 and later support indirect function calls via the [`cmake_language(CALL ...)`](https://cmake.org/cmake/help/latest/command/cmake_language.html#call) construct, which would let us just construct a list of hook functions to call and then call them in order at the right time.

_However_, we currently have CMake 3.16 as our minimum required version of CMake, as this is the newest version shipped by Ubuntu 20.04. Given this we have two options:

- Bump the CMake minimum required version to 3.18 now and add some extra complexity to CI and the package builders to support this.
- Wait until April when Ubuntu 20.04 hits EOL per our platform support policy and we would be bumping the CMake minimum required version to 3.18 anyway.

**Comment by vkalintiris** [comment]
> Bump the CMake minimum required version to 3.18 now and add some extra complexity to CI and the package builders to support this.

I vote for this.

**Comment by Ferroin** [comment]
> > Bump the CMake minimum required version to 3.18 now and add some extra complexity to CI and the package builders to support this.
> 
> I vote for this.

OK, we’ll go with that then, though at this point it’s likely to be January before I get to it (unless I decide to do some work over the holidays).

**Comment by ilyam8** [comment]
Guys, was this addressed? Do we still need this issue?

**Comment by Ferroin** [comment]
No, it was not fully addressed, and yes we do still need to track it.

**Comment by vkalintiris** [comment]
Re-checked the state of this. The blocker from December 2024 is gone, but I'd like us to wait a few more weeks before picking it up, for a reason that only became visible now.

### The original blocker is resolved

The concern was that we needed CMake 3.18 for `cmake_language(CALL ...)` while our minimum was 3.16 (Ubuntu 20.04). Ubuntu 20.04 has since moved to `legacy:` in `.github/data/distros.yml`. Nothing in the actively supported set requires 3.16 anymore.

### CMake floor across supported platforms

Checked every entry under `include:` in `.github/data/distros.yml` against the cmake version the distro actually ships:

| Platform | distro cmake |
| --- | --- |
| **debian 11 (bullseye)** | **3.18.4** |
| centos-stream 9 | 3.20.2 |
| ubuntu 22.04 | 3.22.1 |
| amazonlinux 2023 | 3.22.2 |
| debian 12 | 3.25.1 |
| rockylinux 8 / oraclelinux 8 | 3.26.5 |
| centos-stream 10 | 3.28.2 |
| ubuntu 24.04 | 3.28.3 |
| debian 13, ubuntu 25.10, opensuse 16.0, alpine 3.22, fedora 43 | 3.31.x |
| alpine 3.23, ubuntu 26.04, fedora 44, archlinux, tumbleweed | 4.x |

Two platforms ship older cmake — CentOS 7 (2.8.12 base, 3.17.5 from EPEL) and Amazon Linux 2 (3.13.1 / 3.17.5). Neither is affected by this change: both are already below our current 3.16 minimum, both have carried a bundled CMake in their package builders since netdata/helper-images@0d591a9 ("Add newer CMake to RHEL 7 equivalent package builders"), and both are marked `skip-local-build: true` so no CI source build runs against their distro cmake.

One thing worth stating explicitly so it isn't misread later: the RPM v2 builders now install a pinned CMake 4.1.6 (netdata/helper-images@976ff83, `scripts/install-cmake.sh`). That is a CPack requirement — CPack only emits RPM weak dependencies from CMake 4.1 on — not evidence that we require a modern CMake to build. Everything outside the RPM packaging path is still expected to work with distro-native CMake.

### Why I'd rather wait

Debian 11 gives us exactly 3.18.4 — enough for `cmake_language(CALL ...)`, but with zero headroom. In particular `cmake_language(DEFER CALL ...)` is 3.19 and would be unavailable.

We track `extendedSupport` for Debian (`eol_lts: true`), and Debian 11's ends **2026-08-31**. Once it drops, the floor becomes Debian 12 at 3.25.1.

So rather than bumping to 3.18 now and very likely bumping again shortly after, I'd suggest we revisit this after 2026-08-31 and go straight to 3.20 or higher in a single change.

### Notes for whoever picks this up

- The refactor itself does not need `DEFER`. Deferred calls run at the end of the directory scope, which is too late here: the post-fetch hooks have to run after the centralized `FetchContent_MakeAvailable()` but *before* targets are wired up. For example `packaging/cmake/Modules/NetdataJSONC.cmake` creates the `json-c-compat-link` target that consumers then `add_dependencies()` on. An explicit hook list replayed with `cmake_language(CALL ...)` puts that at the right point.
- Sites needing the declare/make-available split: Corrosion, mimalloc and libdatachannel in the top-level `CMakeLists.txt`, plus json-c, absl, protobuf, yaml, dlib, sentry and ibm_mq under `packaging/cmake/Modules/`.
- The underestimated part is the `netdata_detect_*` macros. They currently decide bundled-vs-system *and* act in one pass; they need splitting into a declare phase and a hook phase.
- Hook ordering matters — protobuf's has to run after absl's.
- Unrelated pre-existing gap noticed while checking this: `packaging/installer/install-required-packages.sh` installs plain `cmake` with no CentOS 7 / Amazon Linux 2 override, which resolves to 2.8.12 on CentOS 7. That path already fails against our current 3.16 minimum. It is rarely hit because kickstart tries native packages and static builds before building from source, but it should be tracked separately.
